linkerd/trace-context/src/propagation/b3.rs

```diff
@@ -0,0 +1,163 @@
+use crate::propagation::try_split_to;
+use crate::{Flags, Id};
+use bytes::Bytes;
+use http::header::HeaderValue;
+use linkerd_error::Error;
+use rand::thread_rng;
+
+use tracing::{trace, warn};
+
+use super::{decode_id_with_padding, get_header_str, Propagation, TraceContext, UnknownFieldId};
+
+static HTTP_TRACE_ID_HEADER: http::header::HeaderName =
+    http::header::HeaderName::from_static("x-b3-traceid");
+static HTTP_SPAN_ID_HEADER: http::header::HeaderName =
+    http::header::HeaderName::from_static("x-b3-spanid");
+static HTTP_SAMPLED_HEADER: http::header::HeaderName =
+    http::header::HeaderName::from_static("x-b3-sampled");
+
+static GRPC_TRACE_HEADER: http::header::HeaderName =
+    http::header::HeaderName::from_static("grpc-trace-bin");
+const GRPC_TRACE_FIELD_TRACE_ID: u8 = 0;
+const GRPC_TRACE_FIELD_SPAN_ID: u8 = 1;
+const GRPC_TRACE_FIELD_TRACE_OPTIONS: u8 = 2;
+
+// This code looks significantly weirder if some of the elements are added using
+// the `vec![]` macro, despite clippy's suggestions otherwise...
+#[allow(clippy::vec_init_then_push)]
+pub fn increment_grpc_span_id<B>(request: &mut http::Request<B>, context: &TraceContext) -> Id {
+    let span_id = Id::new_span_id(&mut thread_rng());
+
+    trace!(message = "incremented span id", %span_id);
+
+    let mut bytes = Vec::<u8>::new();
+
+    // version
+    bytes.push(0);
+
+    // trace id
+    bytes.push(GRPC_TRACE_FIELD_TRACE_ID);
+    bytes.extend(context.trace_id.0.iter());
+
+    // span id
+    bytes.push(GRPC_TRACE_FIELD_SPAN_ID);
+    bytes.extend(span_id.0.iter());
+
+    // trace options
+    bytes.push(GRPC_TRACE_FIELD_TRACE_OPTIONS);
+    bytes.push(context.flags.0);
+
+    let bytes_b64 = base64::encode(&bytes);
+
+    if let Result::Ok(hv) = HeaderValue::from_str(&bytes_b64) {
+        request.headers_mut().insert(&GRPC_TRACE_HEADER, hv);
+    } else {
+        warn!("invalid header: {:?}", &bytes_b64);
+    }
+    span_id
+}
+
+pub fn increment_http_span_id<B>(request: &mut http::Request<B>) -> Id {
+    let span_id = Id::new_span_id(&mut thread_rng());
+
+    trace!("incremented span id: {span_id}");
+
+    let span_str = hex::encode(span_id.as_ref());
+
+    if let Ok(hv) = HeaderValue::from_str(&span_str) {
+        request.headers_mut().insert(&HTTP_SPAN_ID_HEADER, hv);
+    } else {
+        warn!("invalid {HTTP_SPAN_ID_HEADER} header: {span_str:?}");
+    }
+    span_id
+}
+
+pub fn unpack_grpc_trace_context<B>(request: &http::Request<B>) -> Option<TraceContext> {
+    get_header_str(request, &GRPC_TRACE_HEADER)
+        .and_then(|header_str| {
+            base64::decode(header_str)
+                .map_err(|e| warn!("trace header {GRPC_TRACE_HEADER} is not base64 encoded: {e}"))
```

Comment:
```suggestion
                .map_err(|e| warn!("{GRPC_TRACE_HEADER} is not base64 encoded: {e}"))
```

Looking at https://docs.rs/base64/latest/src/base64/decode.rs.html#10-28, I'm skeptical that the error messages we have here are useful to users.